scripts/checkpatch: Limit which feature test macros we prevent

aescolar opened the pull request:
Unfortunately this check as it is today is causing trouble, while not checking too well for what it intended. Let's reduce its scope until a better solution has been found.

Background:
This check intends to ensure coding guidelines Rules A.4 and A.5 are followed, but how it is implemented it does not work well enough.
1. These rules only apply to the kernel and some other parts of the embedded codebase respectively, but this check is performed on the whole tree.
2. This check works under the assumption that any attempt to set these macros in source files is a violation of these rules, while this is not necessary the case, as there are legitimate uses for these. (Specially for _POSIX_C_SOURCE and _XOPEN_SOURCE)

This check also fails to detect these macros being set in cmake files,
so if users are faced with this failure they can trivially bypass it.

Having a CI check which produces too many false positives, while at the same time being very easy to bypass is not a desirable situation as that can result in lack of trust for this type of checks, and an overall tendency to override these CI faults, and overlooking actual violations of these rules by reviewers.

This check was originally added in
b021dece987d691342636dad836d6c9611d510e1

---

References: 
* https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-4-c-standard-library-usage-restrictions-in-zephyr-kernel

Fixes: #67283

---

Edit on 2024/02/16 16:58 due to change of scope of PR as it got change to the approach proposed by @stephanosio

stephanosio reviewed (changes requested):
@aescolar Instead of getting rid of this check altogether, should we consider making it a check for non-standard feature test macros? (i.e. remove `_POSIX_C_SOURCE` and `_XOPEN_SOURCE` from the blacklist and change the error message to something like "do not specify non-standard feature test macros")

It is still useful to prevent the use of non-standard feature test macros like `_GNU_SOURCE` and `_BSD_SOURCE`.

p.s. As for the enforcement of the Rule A.4 and Rule A.5, as previously discussed, we will still need to take a different approach (e.g. making an explicit list of disallowed functions ...), which is beyond the scope of this change and needs to be addressed separately.

aescolar committed:
scripts/checkpatch: Limit which feature test macros we prevent

Unfortunately this check as it is today is causing trouble, while
not checking too well for what it intended. Let's reduce its scope
until a better solution has been found.

Background:
This check intends to ensure coding guidelines Rules A.4 and A.5
are followed, but how it is implemented it does not work well enough.
1. These rules only apply to the kernel and some other parts of the
embedded codebase respectively, but this check is performed on the
whole tree.
2. This check works under the assumption that any attempt to set
these macros in source files is a violation of these rules, while
this is not necessary the case, as there are legitimate uses for these.
(Specially for _POSIX_C_SOURCE and _XOPEN_SOURCE)

This check also fails to detect these macros being set in cmake files,
so if users are faced with this failure they can trivially bypass it.

Having a CI check which produces too many false positives, while
at the same time being very easy to bypass is not a desirable situation
as that can result in lack of trust for this type of checks,
and an overall tendency to override these CI faults,
and overlooking actual violations of these rules by reviewers.

This check was originally added in
b021dece987d691342636dad836d6c9611d510e1

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar commented:
> @aescolar Instead of getting rid of this check altogether, should we consider making it a check for non-standard feature test macros? 

@stephanosio  PR changed as per your suggestion.

MaureenHelm reviewed (approved):
@keith-zephyr can you take another look please?

keith-packard reviewed (commented):
I think we need to continue to exclude `_POSIX_C_SOURCE` and `_XOPEN_SOURCE` from files in whatever is defined as the Zephyr "core"? Could we add another check to look for those based on the pathname of the file being patched?

cfriedt commented:
I would have to agree with Keith in this case.

cfriedt reviewed (dismissed):
My opinion is that this does not fix the script in a way that aligns with the issue. I.e. I think we should continue to enforce that _POSIX_C_SOURCE and _XOPEN_SOURCE are not declared based on path, to keep POSIX out of the "core" part of Zephyr.

I also feel that merging this would circumvent part of the process of another PR being reviewed.

aescolar commented:
@keith-packard @cfriedt 
Limiting this check to only some paths would reduce the number of false positives, but the check would still not be correct.
The thing is, this check is not finding violations of rules A.4/5,
instead it is reacting to one of the possible things a user may be using to enable breaking that rule, but which may also be used for legitimate reasons; while at the same time being blind to a user breaking it with other enablers.

As mentioned by @stephanosio, a different approach for this check is needed, which could be
similar to the check for MISRA rule 21.2
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/coccinelle/symbols.txt

cfriedt commented:
> instead it is reacting to one of the possible things a user may be using to enable breaking that rule, but which may also be used for legitimate reasons;

@aescolar - can you list the legitimate reasons?

aescolar commented:
@cfriedt 

> @aescolar - can you list the legitimate reasons?

At least:
* Any core/kernel or on tree component which uses an API which is not exposed by default by all C libraries that component may be built with but which is allowed by coding guidelines A.4/A.5.
  * This includes code using new functions added to that list like for ex. in #68414
* Code only meant to run in the "native" targets (this includes code in the arch/soc/boards, drivers and backends specific for these & test code specific for these)
* Components designed under the understanding that they have access to more than that basic C standard API, both those which may provide some of those APIs (for ex. networking subsystem, libC, POSIX OS API abstraction layer), where parts of those components may rely on functionality provided by other parts of the same component, or where part of that component may have a hard dependency on another of these components; their tests or samples using those components.

aescolar commented:
> To clarify, what are the legitimate reasons we would want to allow `_POSIX_C_SOURCE` definitions to be used by "core" parts of Zephyr such as the kernel?
> 
> You mentioned `gmtime_r()`. Are there other such functions? If so, how many?

For the kernel itself only `strnlen()` would be allowed today, though others may be added in the future to rule A.4.
In that point I also referred to other in-tree embedded components; for those is for which those ones also listed today in, or which may be added later to, A.5 would apply.

cfriedt commented:
> > To clarify, what are the legitimate reasons we would want to allow `_POSIX_C_SOURCE` definitions to be used by "core" parts of Zephyr such as the kernel?
> > You mentioned `gmtime_r()`. Are there other such functions? If so, how many?
> 
> For the kernel itself only `strnlen()` would be allowed today, though others may be added in the future to rule A.4. In that point I also referred to other in-tree embedded components; for those is for which those ones also listed today in, or which may be added later to, A.5 would apply.

Taking a step back, it occurred to me that what we're observing here is the result of what some people call a "layering violation". These are not always super obvious, but it's easy to visualize.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/983494/2deecd09-d13d-4c40-be11-c82d7eac52d4)

Missing from the graphic above is the C library, which can probably be split into OS-dependent portions of the C library and OS-independent portions of the C library. Apologies - there may be a more conventional term for that, but I do not know it off-hand - so please bear with my non-standard terminology. For clarity, "C library" refers to the mandatory requirements for ISO C, not POSIX extensions to the ISO C standard.

Everything above hardware can rely on the OS-independent portion of the C library, and everything above the kernel can rely on OS-dependent portions of the C library. That makes a nice, linear directed acyclic dependency graph (DAG).

Generally, dependencies should only go downward and should not have cycles. There are some special circumstances where a reverse dependency can occur (e.g. OSAL - for the BSP to support Zephyr, for example) but that should not introduce cycles (and therefore still have a DAG).

Currently, we have voluntarily injected a dependency cycle into our "layer cake".

We specifically depend on the 3 POSIX APIs below probably where we should not be. A dependency cycle is one possible result of such a layering violation.

I would say a far better solution that is ultimately much more scalable and requires zero special treatment of anything is to simply create kernel-level APIs for the 3 functions below. They are mostly trivial - integer or text manipulation.

* `strnlen()` -> `k_strnlen()`
* `gmtime_r()` -> `k_gmtime_r()`
* `strtok_r()` -> `k_strtok_r()`

The `k_` prefix is kind of arbitrary - we could use `z_`, but I know @nashif  wants to get rid of that. `k_` is public, which is why I suggested it.

While it's true that defining `_POSIX_C_SOURCE` would expose the 3 functions above to Zephyr libraries and applications, it would likely also expose the entire [1191 system interfaces](https://pubs.opengroup.org/onlinepubs/9699919799/idx/functions.html) in circumstances where they should arguably not be exposed.

That number might even be outdated at this point.

It is very clear that we are overemphasizing the "weight" of the 3 text and integer manipulation functions above.

To me, it's a no-brainer to simply remove the dependency cycle.

aescolar commented:
@cfriedt
Those 3 functions and many other are not OS related, but provided by the basic C library
without any OS awareness.
Many of those have, over the years, been added to the ISO C standards (for ex. gmtime_r in C23).

But C libraries, when built with the strict iso flag set (as we do) tend to not expose these
function prototypes unless the required iso standard or one of the corresponding
feature test macros is set.
Even though we could define our own shims in Zephyr for these 3 functions, that would
break users expectations, and require more code porting.

But even if we did that, there would still be other legitimate uses and need for these feature
test macros in the tree.

The want to remove this checkpatch rule boils down to just:
* There is no rule against setting these feature test macros in Zephyr, and as such this
  checkpatch check is preventing something which is not fordidden or incorrect.
* The checkpatch check was intended to enforce another rule, but implementing the 
  check in this way was simpler and seemed good enough. Unfortunately the check has
  proven to be problematic: It causes a lot of false positives, while it does not
  actually detect rules' A.4/A.5 violations in many cases and is trivial to bypass.

I'd request you reconsider your rejection of this PR, as we keep getting this checkpatch
check triggered all too often, which confuses developers and causes extra work for
too many people, so I would prefer the check to not be there anymore.

cfriedt commented:
I would suggest the following 

1. Mitigate the layering violations with `strnlen()`, `strtok_r()`, and `gmtime_r()` by prefixing the Zephyr implementations with a `k_`, adding them to `lib/os`, and declaring them in `kernel.h`. This eliminates a dependency cycle in the DAG which should probably have been avoided in the first place. It also circumvents all areas where `_POSIX_C_SOURCE` would need to be defined currently (outside of the posix arch). 

2. Add an exception to `checkpatch.pl` for native paths such as `arch/posix`, `soc/posix`, etc. Keep existing CI checks in place so that `_POSIX_SOURCE` and other reserved macros are not ad-hoc sprinkled throughout the codebase with possibly conflicting versions.

3. Adopt #67132 as a conventional, consistent, and user-configurable way to specify POSIX application conformance levels.

aescolar commented:
@cfriedt

> Strnlen and strtok_r are OS independent (correct) but again, there very much is a layering violation at the interface level.

Not really. In that drawing you pasted you present the POSIX functionality as if it were one component which uses the kernel, when in reality POSIX.1-2017 is a standard which specifies many parts of an OS and the C library.
Wrt. the C library, as you yourself stated, there is a significant set of functions which are OS agnostic,
and which can even be used baremetal.

You should note that in [the list you linked](https://pubs.opengroup.org/onlinepubs/9699919799/idx/functions.html) many of those functions are indeed part of the ISO C standard, and many of those do not require or depend on in any way an OS kernel, or in Zephyr's case, the code behind CONFIG_POSIX_API either.
The "arrow" in those would simply point towards the C library part which sits directly on the metal.

bjarki-trackunit commented:
Hi, adding my opinion here from a user's perspective:

I expect all "standard" (notice the quotes) C functions placed in the coding guidelines, to be available to the kernel/application, regardless of which compiler/board/soc is used. `strlen()` and `strnlen()` are placed in the same list, and should be exactly as available as each other.

However, I do see that we are conflicting with POSIX by defining `strnlen()` as a standard C function, which is expected to be provided by the C compiler, rather than POSIX.

So, I believe the appropriate fix is to take ownership of `strnlen()` in the kernel, and namespace it appropriately (`k_strnlen()` is a good option) and provide an implementation of it to be used within the kernel.

This allows POSIX compliant applications to include and use the POSIX implementation, and the kernel/subsystems to use the kernel implementation.

If/when strnlen() becomes part of ISO C, we deprecate `k_strnlen()`, and simply use strnlen() which is now no longer conflicting with POSIX

aescolar commented:
Thanks for the input @bjarki-trackunit

> I expect all "standard" (notice the quotes) C functions..

We are trying to close on that topic in https://github.com/zephyrproject-rtos/zephyr/issues/68278,
but in very short: quite a few people have that expectation today, but doing so is unfortunately 
not as easy as one could hope, but if that expectation is not met, I hope to clarify the wording in rules A.4 and A.5.

Note that for strnlen there is, in practice, just the implementation provided by the C library.
There is no incompatibility between strnlen() implementations.
The issue users face is just that its prototype is not exposed by default.

Note that even if we were to ensure all "Zephyr provided librarires" expose this prototype by
default, any Zephyr code which aimed at being properly portable to other C libraries
should still set _POSIX_C_SOURCE if it uses strnlen(), so that it works with other
standards compliant libraries. So this checkpatch check as it is today, is actully preventing
writing code in Zephyr that will work with users propietary C libraries out of the box.

I don't think defining k_strnlen() et al. would be a good solution because developers will still
expect strnlen() and we are already using it today. So we'd just add the overhead of having a shim
or duplicate implementation only to avoid setting a feature test macro.
It also feels like defining our own k_strnlen(), ultimately, just because we have an incorrect checkpatch
check is not the way to go.

aescolar commented:
Hi @keith-zephyr, unfortunately this PR has been blocked now for more than a month.
As assignee to the PR and owner of the process, I would request you to consider dismissing his review so that this PR can progress:
Even if all changes requested by @cfriedt were agreed, and all applied, the issue this PR is solving would not be solved (this checkpatch check in question would remain incorrect and would keep causing problems).

cfriedt commented:
I'm confused... It really does not seem like anything has been done to address the concerns raised by @keith-packard, myself, and @bjarki-trackunit .

keith-zephyr commented:
> I'm confused... It really does not seem like anything has been done to address the concerns raised by @keith-packard, myself, and @bjarki-trackunit .
> 
> One concern was to limit checking `_POSIX_C_SOURCE` and `_XOPEN_SOURCE` by path.

@cfriedt : I'm in agreement with @aescolar, that the existing checks for `_POSIX_C_SOURCE` and `_XOPEN_SOURCE`  aren't preventing violations of rule A.4 and A.5.  So I would like this PR to move forward.
